lib/opam_solve.ml

```diff
@@ -251,6 +254,65 @@ module Make_solver (Context : OPAM_MONOREPO_CONTEXT) :
       rolemap []
     |> List.filter_map ~f:Solver.package_name
 
+  let no_matching_versions diagnostics =
+    let rolemap = Solver.diagnostics_rolemap diagnostics in
+    Pkg_map.fold
+      (fun pkg component acc ->
+        match Solver.Diagnostics.Component.selected_impl component with
+        | Some _ -> acc
+        | None ->
+            (* short-circuit skip of fold *)
+            let ( let* ) a f = match a with Some a -> f a | None -> acc in
+            let* pkg_name = Solver.package_name pkg in
+            let notes = Solver.Diagnostics.Component.notes component in
+            let* _, version_restriction =
+              List.find_map
+                ~f:(function
+                  | UserRequested restriction -> Some restriction
+                  | Restricts (_other_role, _impl, restrictions) -> (
+                      match restrictions with
+                      | [] -> None
+                      | restriction :: _ -> Some restriction)
+                  | _ -> None)
+                notes
+              |> Option.map ~f:Solver.formula
+            in
+            let rejects, _reason =
+              Solver.Diagnostics.Component.rejects component
+            in
+            (* find only the model rejections that is those that we have rejected as e.g. not building with dune *)
+            let model_rejected =
+              List.filter_map
```

Comment:
I'm not sure it's worth filtering those out first, I think we can do the check in a single pass.